calendar backup not working

I've enabled daily contacts and calendar backup.

Backup of contacts is working for a long time now and it's backup up all contacts to a single .vcf file in .Contacts-Backup on the nextcloud server home of the user (I would like to specify the folder but that's another topic :) )

However the newly introduced calendar backup doesn't do anything. Or where should I find the backup files?

It's Nextcloud files for android 3.18.0 on Android 11.

Best
Henning

From the above log:

```java
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: Work [ id=a948a45c-10cf-49c1-bb5d-0fe9bda45e1c, tags={ name:immediate_calendar_backup <account id> timestamp:1661764601420, *, user:<user id>, com.nextcloud.client.jobs.CalendarBackupWork } ] failed because it threw an exception/error
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: java.util.concurrent.ExecutionException: net.fortuna.ical4j.model.ValidationException: Calendar must contain at least one component
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:516)
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:475)
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:311)
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at java.lang.Thread.run(Thread.java:920)
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: Caused by: net.fortuna.ical4j.model.ValidationException: Calendar must contain at least one component
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at net.fortuna.ical4j.model.Calendar.validate(Calendar.java:287)
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at net.fortuna.ical4j.model.Calendar.validate(Calendar.java:257)
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at net.fortuna.ical4j.data.CalendarOutputter.output(CalendarOutputter.java:96)
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at net.fortuna.ical4j.data.CalendarOutputter.output(CalendarOutputter.java:83)
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at third_parties.sufficientlysecure.SaveCalendar.start(SaveCalendar.java:193)
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at com.nextcloud.client.jobs.CalendarBackupWork.doWork(CalendarBackupWork.kt:69)
08-29 11:16:41.606 14654 14672 E WM-WorkerWrapper: 	at androidx.work.Worker$1.run(Worker.java:86)
```
I have no idea what `Calendar must contain at least one component`, I'll try to figure it out.

Does not work here either
My main calendar (size 1.5 MB) is still missing

Pixel4
Android 11
Nextcloud 3.22.1

`10-24 05:19:16.715 I/ActivityTaskManager(1536): START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=com.nextcloud.client/com.owncloud.android.ui.activity.FileDisplayActivity bnds=[838,990][1038,1314]} from uid 10237
10-24 05:19:17.121 D/ArtManagerInternalImpl(1536): /data/misc/iorapd/com.nextcloud.client/30220190/com.owncloud.android.ui.activity.FileDisplayActivity/compiled_traces/compiled_trace.pb doesn't exist
10-24 05:19:17.123 I/ActivityTaskManager(1536): Displayed com.nextcloud.client/com.owncloud.android.ui.activity.FileDisplayActivity: +398ms
10-24 05:19:17.208 I/GoogleInputMethodService(22332): GoogleInputMethodService.onStartInput():1960 onStartInput(EditorInfo{inputType=0x0(NULL) imeOptions=0x0 privateImeOptions=null actionName=UNSPECIFIED actionLabel=null actionId=0 initialSelStart=-1 initialSelEnd=-1 initialCapsMode=0x0 hintText=null label=null packageName=com.nextcloud.client fieldId=-1 fieldName=null extras=null}, false)
10-24 05:19:17.747 I/UpdateOCVersionOperation(15698): Check for update of Nextcloud server version at https://xxx.com/remote.php/dav/files/xxx: Operation finished with HTTP status code -1 (success)
10-24 05:19:17.797 I/UpdateOCVersionOperation(15698): Check for update of Nextcloud server version at https://xxx.com/remote.php/dav/files/xxx: Operation finished with HTTP status code -1 (success)
10-24 05:19:21.594 I/ActivityTaskManager(1536): START u0 {cmp=com.nextcloud.client/com.owncloud.android.ui.activity.SettingsActivity} from uid 10284
10-24 05:19:21.728 I/ActivityTaskManager(1536): Displayed com.nextcloud.client/com.owncloud.android.ui.activity.SettingsActivity: +133ms
10-24 05:19:21.766 I/GoogleInputMethodService(22332): GoogleInputMethodService.onStartInput():1960 onStartInput(EditorInfo{inputType=0x0(NULL) imeOptions=0x0 privateImeOptions=null actionName=UNSPECIFIED actionLabel=null actionId=0 initialSelStart=-1 initialSelEnd=-1 initialCapsMode=0x0 hintText=null label=null packageName=com.nextcloud.client fieldId=-1 fieldName=null extras=null}, false)
10-24 05:19:23.855 I/ActivityTaskManager(1536): START u0 {cmp=com.nextcloud.client/com.owncloud.android.ui.activity.ContactsPreferenceActivity (has extras)} from uid 10284
10-24 05:19:23.985 I/GoogleInputMethodService(22332): GoogleInputMethodService.onStartInput():1960 onStartInput(EditorInfo{inputType=0x0(NULL) imeOptions=0x0 privateImeOptions=null actionName=UNSPECIFIED actionLabel=null actionId=0 initialSelStart=-1 initialSelEnd=-1 initialCapsMode=0x0 hintText=null label=null packageName=com.nextcloud.client fieldId=-1 fieldName=null extras=null}, false)
10-24 05:19:25.884 I/ICS_SaveCalendar(15698): Save id 1 to file /data/user/0/com.nextcloud.client/cache/Test_2022-10-24_05-19-25.ics
10-24 05:19:26.018 I/ICS_SaveCalendar(15698): Save id 2 to file /data/user/0/com.nextcloud.client/cache/Moon/Astro_2022-10-24_05-19-26.ics
10-24 05:19:26.086 E/WM-WorkerWrapper(15698): Work [ id=a05649ef-6c9a-422c-be4e-c794c8afd7a7, tags={ name:immediate_calendar_backup xxx@xxx.com, *, user:xxx@xxx.com, timestamp:1666581565757, com.nextcloud.client.jobs.CalendarBackupWork } ] failed because it threw an exception/error
10-24 05:19:26.086 E/WM-WorkerWrapper(15698): java.util.concurrent.ExecutionException: java.io.FileNotFoundException: /data/user/0/com.nextcloud.client/cache/Moon/Astro_2022-10-24_05-19-26.ics: open failed: ENOENT (No such file or directory)
10-24 05:19:26.086 E/WM-WorkerWrapper(15698): Caused by: java.io.FileNotFoundException: /data/user/0/com.nextcloud.client/cache/Moon/Astro_2022-10-24_05-19-26.ics: open failed: ENOENT (No such file or directory)
10-24 05:19:26.086 E/WM-WorkerWrapper(15698): 	at com.nextcloud.client.jobs.CalendarBackupWork.doWork(CalendarBackupWork.kt:68)
10-24 05:19:26.087 I/WM-WorkerWrapper(15698): Worker result FAILURE for Work [ id=a05649ef-6c9a-422c-be4e-c794c8afd7a7, tags={ name:immediate_calendar_backup xxx@xxx.com, *, user:xxx@xxx.com, timestamp:1666581565757, com.nextcloud.client.jobs.CalendarBackupWork } ]
10-24 05:19:26.222 D/NotificationService(1536): 0|com.nextcloud.client|1000000|null|10284: granting content://settings/system/notification_sound
10-24 05:19:26.226 D/NotificationService(1536): 0|com.nextcloud.client|1000000|null|10284: granting content://settings/system/notification_sound
10-24 05:19:26.247 D/InterruptionStateProvider(2114): No bubble up: not allowed to bubble: 0|com.nextcloud.client|1000000|null|10284
10-24 05:19:26.248 D/InterruptionStateProvider(2114): No heads up: unimportant notification: 0|com.nextcloud.client|1000000|null|10284
10-24 05:19:26.255 D/InterruptionStateProvider(2114): No bubble up: not allowed to bubble: 0|com.nextcloud.client|1000000|null|10284
10-24 05:19:26.255 D/InterruptionStateProvider(2114): No heads up: unimportant notification: 0|com.nextcloud.client|1000000|null|10284
10-24 05:19:29.051 I/UploadFileOperation(15698): Upload of /data/user/0/com.nextcloud.client/cache/Test_2022-10-24_05-19-25.ics to /.Calendar-Backup/Test_2022-10-24_05-19-25.ics: Operation finished with HTTP status code 201 (success)
10-24 05:19:35.732 I/GoogleInputMethodService(22332): GoogleInputMethodService.onStartInput():1960 onStartInput(EditorInfo{inputType=0x0(NULL) imeOptions=0x0 privateImeOptions=null actionName=UNSPECIFIED actionLabel=null actionId=0 initialSelStart=-1 initialSelEnd=-1 initialCapsMode=0x0 hintText=null label=null packageName=com.nextcloud.client fieldId=-1 fieldName=null extras=null}, false)
10-24 05:19:36.849 W/ActivityTaskManager(1536): Duplicate finish request for r=ActivityRecord{1f767f9 u0 com.nextcloud.client/com.owncloud.android.ui.activity.SettingsActivity t1648 f}}
10-24 05:19:37.002 I/GoogleInputMethodService(22332): GoogleInputMethodService.onStartInput():1960 onStartInput(EditorInfo{inputType=0x0(NULL) imeOptions=0x0 privateImeOptions=null actionName=UNSPECIFIED actionLabel=null actionId=0 initialSelStart=-1 initialSelEnd=-1 initialCapsMode=0x0 hintText=null label=null packageName=com.nextcloud.client fieldId=-1 fieldName=null extras=null}, false)
10-24 05:19:37.716 I/UpdateOCVersionOperation(15698): Check for update of Nextcloud server version at https://xxx.com/remote.php/dav/files/xxx: Operation finished with HTTP status code -1 (success)`

--- a/app/src/main/java/com/nextcloud/client/jobs/CalendarBackupWork.kt
+++ b/app/src/main/java/com/nextcloud/client/jobs/CalendarBackupWork.kt
@@ -59,7 +59,9 @@ class CalendarBackupWork(
 
         val force = inputData.getBoolean(FORCE, false)
         if (force || lastExecution + JOB_INTERVAL_MS < Calendar.getInstance().timeInMillis) {
-            AndroidCalendar.loadAll(contentResolver).forEach { calendar ->
+            val calendars = AndroidCalendar.loadAll(contentResolver)
+            Log_OC.d(TAG, "Saving ${calendars.size} calendars")
+            calendars.forEach { calendar ->
                 SaveCalendar(
                     applicationContext,
                     calendar,

--- a/app/src/main/java/third_parties/sufficientlysecure/AndroidCalendar.java
+++ b/app/src/main/java/third_parties/sufficientlysecure/AndroidCalendar.java
@@ -101,7 +101,7 @@ public class AndroidCalendar {
             calendar.mTimezone = getString(cur, Calendars.CALENDAR_TIME_ZONE);
 
             final String[] args = new String[]{calendar.mIdStr};
-            Cursor eventsCur = resolver.query(Events.CONTENT_URI, CAL_ID_COLS, CAL_ID_WHERE, args, null);
+            Cursor eventsCur = resolver.query(Events.CONTENT_URI, CAL_ID_COLS, CAL_ID_WHERE + " AND deleted=0", args, null);
             calendar.mNumEntries = eventsCur.getCount();
             eventsCur.close();
             calendars.add(calendar);

--- a/app/src/main/java/third_parties/sufficientlysecure/ProcessVEvent.java
+++ b/app/src/main/java/third_parties/sufficientlysecure/ProcessVEvent.java
@@ -433,18 +433,24 @@ public class ProcessVEvent {
             //        - Check the calendars max number of alarms
             if (t.getDateTime() != null) {
                 alarmMs = t.getDateTime().getTime(); // Absolute
-            } else if (t.getDuration() != null && t.getDuration().isNegative()) {
+            } else if (t.getDuration() != null && t.getDuration().isNegative()) { //alarm trigger before start of event
                 Related rel = (Related) t.getParameter(Parameter.RELATED);
                 if (rel != null && rel == Related.END) {
                     alarmStartMs = e.getEndDate().getDate().getTime();
                 }
-                alarmMs = alarmStartMs - durationToMs(t.getDuration()); // Relative
+                alarmMs = alarmStartMs - durationToMs(t.getDuration()); // Relative "-"
+            } else if (t.getDuration() != null && !t.getDuration().isNegative()) { //alarm trigger after start of event
+                Related rel = (Related) t.getParameter(Parameter.RELATED);
+                if (rel != null && rel == Related.END) {
+                    alarmStartMs = e.getEndDate().getDate().getTime();
+                }
+                alarmMs = alarmStartMs + durationToMs(t.getDuration()); // Relative "+"
             } else {
                 continue;
             }
 
             int reminder = (int) ((startMs - alarmMs) / DateUtils.MINUTE_IN_MILLIS);
-            if (reminder >= 0 && !reminders.contains(reminder)) {
+            if (!reminders.contains(reminder)) {
                 reminders.add(reminder);
             }
         }
@@ -520,7 +526,7 @@ public class ProcessVEvent {
     }
 
     private Cursor queryEvents(ContentResolver resolver, StringBuilder b, List<String> argsList) {
-        final String where = b.toString();
+        final String where = b.toString() + " AND deleted=0";
         final String[] args = argsList.toArray(new String[argsList.size()]);
         return resolver.query(Events.CONTENT_URI, EVENT_QUERY_COLUMNS, where, args, null);
     }

--- a/app/src/main/java/third_parties/sufficientlysecure/SaveCalendar.java
+++ b/app/src/main/java/third_parties/sufficientlysecure/SaveCalendar.java
@@ -203,7 +203,7 @@ public class SaveCalendar {
 
             upload(fileName);
         } else {
-            Log_OC.w(TAG, "Calendar has " + selectedCal.mIdStr + "no components");
+            Log_OC.w(TAG, "Calendar '" + selectedCal.mIdStr + "' has no components");
         }
     }
 

--- a/app/src/main/java/third_parties/sufficientlysecure/SaveCalendar.java
+++ b/app/src/main/java/third_parties/sufficientlysecure/SaveCalendar.java
@@ -190,17 +190,21 @@ public class SaveCalendar {
             cal.getComponents().add(v);
         }
 
-        new CalendarOutputter().output(cal, new FileOutputStream(fileName));
+        if (!cal.getComponents().isEmpty()) {
+            new CalendarOutputter().output(cal, new FileOutputStream(fileName));
 
-        Resources res = activity.getResources();
-        String msg = res.getQuantityString(R.plurals.wrote_n_events_to, events.size(), events.size(), file);
-        if (numberOfCreatedUids > 0) {
-            msg += "\n" + res.getQuantityString(R.plurals.created_n_uids_to, numberOfCreatedUids, numberOfCreatedUids);
+            Resources res = activity.getResources();
+            String msg = res.getQuantityString(R.plurals.wrote_n_events_to, events.size(), events.size(), file);
+            if (numberOfCreatedUids > 0) {
+                msg += "\n" + res.getQuantityString(R.plurals.created_n_uids_to, numberOfCreatedUids, numberOfCreatedUids);
+            }
+
+            // TODO replace DisplayUtils.showSnackMessage(activity, msg);
+
+            upload(fileName);
+        } else {
+            Log_OC.w(TAG, "Calendar has " + selectedCal.mIdStr + "no components");
         }
-
-        // TODO replace DisplayUtils.showSnackMessage(activity, msg);
-
-        upload(fileName);
     }
 
     private int ensureUids(Context activity, ContentResolver resolver, AndroidCalendar cal) {

